src/server/hll_family.cc

```diff
@@ -75,26 +75,30 @@ OpResult<int> AddToHll(const OpArgs& op_args, string_view key, CmdArgList values
   RETURN_ON_BAD_STATUS(op_res);
   auto& res = *op_res;
   if (res.is_new) {
-    hll.resize(getDenseHllSize());
-    createDenseHll(StringToHllPtr(hll));
+    hll.resize(getSparseHllInitSize());
+    initSparseHll(StringToHllPtr(hll));
   } else if (res.it->second.ObjType() != OBJ_STRING) {
     return OpStatus::WRONG_TYPE;
   } else {
     res.it->second.GetString(&hll);
-    ConvertToDenseIfNeeded(&hll);
   }
 
   int updated = 0;
+  sds hll_sds = sdsnewlen(hll.data(), hll.size());
+  sds* hll_ptr = &hll_sds;
   for (const auto& value : values) {
-    int added = pfadd(StringToHllPtr(hll), (unsigned char*)value.data(), value.size());
+    // pfadd might resize the input string. A referance to string.data() makes it unnecessarily
+    // complicated
+    int added = pfadd(hll_ptr, (unsigned char*)value.data(), value.size());
```

Comment:
Spent lots of time here. I finally decided not to use HllPtr because Sparse operation expands and reallocates string data. It's not easy to correctly update std::string object  in C.

A sds object is simple and fits the following logic well. 
Some overhead: Copying string to SDS, a reallocating on each sparse insertion, copy SDS back to string.

Reply:
I agree. Indeed sds strings don't play nicely with Dragonfly strings, as they have incompatible layouts.
It looks like your current solution allocates for each additional item in the HLL. Can we make it so that it will only allocate for each additional item in an HLL *if the HLL is a sparse set*? To my understanding, once it becomes dense, there shouldn't be a need to do any further reallocations, because it is of a fixed size. So that could be a nice optimization (and a good trade-off, because the allocations will be on small sizes anyway)
What do you think?